server/app/src/test/java/com/google/heartrate/wearos/app/gatt/heartrate/characteristics/HeartRateGattCharacteristicManagerTest.java

```diff
@@ -0,0 +1,159 @@
+        assertNotThrows(() -> currentHeartRateMeasurementValue[0] = characteristic
+                .getHeartRateMeasurementValue());
+        assertNotThrows(() -> currentExpendedEnergyValue[0] = characteristic
+                .getExpandedEnergyValue());
+
+        assertEquals(heartRateValue, currentHeartRateMeasurementValue[0]);
+        assertEquals(expendedEnergyValue, currentExpendedEnergyValue[0]);
+    }
+
+    /**
+     * Assent set values to given Heart Rate Characteristic fails with GattException.
+     *
+     * @param characteristic      characteristic to set value
+     * @param heartRateValue      heart rate value to set
+     * @param expendedEnergyValue expended energy value to set
+     */
+    public static void assertSetValuesFailed(HeartRateMeasurementCharacteristic characteristic,
+                                             int heartRateValue, int expendedEnergyValue) {
+        assertThrows(GattException.class, () -> characteristic
+                .setHeartRateCharacteristicValue(heartRateValue, expendedEnergyValue));
+    }
+
+    /**
+     * Assent failed set values to given Heart Rate Characteristic not change stored values.
+     *
+     * @param characteristic      characteristic to set value
+     * @param heartRateValue      heart rate value to set
+     * @param expendedEnergyValue expended energy value to set
+     */
+    public static void assertValuesNotChangedAfterFailedSet(HeartRateMeasurementCharacteristic characteristic,
+                                                            int heartRateValue, int expendedEnergyValue) {
+        final int[] beforeSetHeartRateMeasurementValue = new int[1];
+        final int[] beforeSetExpendedEnergyValue = new int[1];
+
+        final int[] currentHeartRateMeasurementValue = new int[1];
+        final int[] currentExpendedEnergyValue = new int[1];
+
+        assertNotThrows(() -> beforeSetHeartRateMeasurementValue[0] = characteristic
+                .getHeartRateMeasurementValue());
+        assertNotThrows(() -> beforeSetExpendedEnergyValue[0] = characteristic
+                .getExpandedEnergyValue());
+
+        assertSetValuesFailed(characteristic, heartRateValue, expendedEnergyValue);
+
+        assertNotThrows(() -> currentHeartRateMeasurementValue[0] = characteristic
+                .getHeartRateMeasurementValue());
+        assertNotThrows(() -> currentExpendedEnergyValue[0] = characteristic
+                .getExpandedEnergyValue());
+
+        assertEquals(beforeSetHeartRateMeasurementValue[0], currentHeartRateMeasurementValue[0]);
+        assertEquals(beforeSetExpendedEnergyValue[0], currentExpendedEnergyValue[0]);
+    }
+
+    @Test
+    public void setHeartRateMeasurementValueOutUInt16AfterUInt16Test() {
+        int heartRateUInt16Value = CharacteristicsArgumentProvider.provide(IN_UINT16)[0];
+        int expendedEnergyUInt16Value = CharacteristicsArgumentProvider.provide(IN_UINT16)[1];
```

Comment:
I think just using the direct values would be easier to read here. Don't worry too much about copying code around in tests :).

Reply:
Fixed in [commit](https://github.com/googleinterns/heartrate-bt-wear/pull/1/commits/93a29ed3ea6d7a2bf9235ec829a41d0e44889a43)